src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.InFlightSolution.cs

```diff
@@ -51,10 +55,31 @@ public sealed class InFlightSolution
                 Checksum solutionChecksum,
                 Func<CancellationToken, Task<Solution>> computeDisconnectedSolutionAsync)
             {
+                Contract.ThrowIfFalse(workspace._gate.CurrentCount == 0);
+
                 _workspace = workspace;
                 SolutionChecksum = solutionChecksum;
 
-                _disconnectedSolutionTask = computeDisconnectedSolutionAsync(_cancellationTokenSource.Token);
+                // Grab the cancellation token up front while we know we are holding the lock and mutating state.  We
+                // don't want to potentially grab it at some undefined point at the future when this type has already
+                // had its in-flight-count reduced to 0 and thus had our CTS be disposed.
+                //
+                // Also kick this off in a dedicated task.  The state mutation updating of this InFlightSolution and the
+                // cache state in RemoteWorkspace must always run fully to completion without issue.  We don't want
+                // anything we call in the constructor to possibly prevent the constructor from running to completion.
+                var cancellationToken = this.CancellationToken;
+                _disconnectedSolutionTask = Task.Run(() => computeDisconnectedSolutionAsync(cancellationToken), cancellationToken);
```

Comment:
this is to be hyper paranoid, but it is legitimately safer.  Synchronously calling into a callback which runs abitrary code within this constructor is inherrently unsafe.  If anything throws synchronously in that callback then this constructor won't finish, and the caller won't properly update cache state.